File: Source/ChannelStuff/controllers/ChannelPlayerTrigger.cs



using System;
using Celeste.Mod.Entities;
using Celeste.Mod.auspicioushelper;
using Microsoft.Xna.Framework;

namespace Celeste.Mod.auspicioushelper;

[CustomEntity("auspicioushelper/ChannelPlayerTrigger")]
public class ChannelPlayerTrigger:Trigger{
  string channel {get; set;}
  public enum Op{
    xor, and, or,
    set, max, min,
    add,
  }
  public Op op;
  public int value;
  bool activateOnEnter=false;
  bool activateOnleave=false;
  bool onlyOnce;

  public ChannelPlayerTrigger(EntityData data, Vector2 offset):base(data, offset){
    onlyOnce = data.Bool("only_once",false);
    channel = data.Attr("channel","");
    value = data.Int("value",1);
    switch(data.Attr("action")){
      case "dash":
        Add(new DashListener((Vector2 d)=>{
          if(PlayerIsInside)activate();
        }));
        break;
      case "jump":
        Add(new JumpListener((int t)=>{
          if(PlayerIsInside)activate();
        }));
        break;
      case "enter":
        activateOnEnter=true; break;
      case "leave":
        activateOnleave=true; break;

      default: DebugConsole.Write("Unknown action"+data.Attr("action")); break;
    }
    op = data.Attr("op","") switch {
      "xor"=>Op.xor,
      "and"=>Op.and,
      "or"=>Op.or,
      "max"=>Op.max,
      "min"=>Op.min,
      "add"=>Op.add,
      _=>Op.set
    };
  }
  public void activate(){
    int oldval = ChannelState.readChannel(channel);
    //DebugConsole.Write(op.ToString());
    ChannelState.SetChannel(channel, op switch {
      Op.set => value,
      Op.xor => oldval ^ value,
      Op.and => value & oldval,
      Op.or => value | oldval,
      Op.add => oldval+value,
      Op.max => Math.Max(oldval,value),
      Op.min => Math.Min(value, oldval),
      _=>oldval
    });
    if(onlyOnce) RemoveSelf();
  }
  public override void OnEnter(Player player){
    base.OnEnter(player);
    if(activateOnEnter) activate();
  }
  public override void OnLeave(Player player){
    base.OnLeave(player);
    if(activateOnleave) activate();
  }
}

File: Loenn/triggers/ChannelPlayerTrigger.lua
local mods = require("mods")

local trigger = {}

trigger.name = "auspicioushelper/ChannelPlayerTrigger"
trigger.triggerText = function(room, self)
    if self.op == "set" then return string.format("set channel \"%s\" to %s", self.channel, self.value) end
    if self.op == "max" or self.op == "min" then
        return string.format("%s(channel \"%s\", %s)", self.op, self.channel, self.value)
    end
    
    return string.format("channel \"%s\" %s %s", self.channel, self.op, self.value)
end
local actions = {"jump","dash","enter","leave"}
local ops = {"xor", "and", "or", "set", "max", "min", "add"}
trigger.placements = {
    name = "main",
    data = {
      channel = "",
      value = 1,
      op = "set",
      action = "dash",
      only_once = false
    }
}
trigger.fieldInformation = {
  op = {
    options=ops
  },
  action = {
    options = actions
  },
  value = {
    fieldType="integer"
  }
}
return trigger